packages/babel-node/src/babel-node.ts

```diff
@@ -127,20 +29,124 @@ if (process.env.BABEL_8_BREAKING) {
+  if (process.env.BABEL_8_BREAKING) {
+    const babelOptions = new Set([]);
+    const babelOptionsWithValue = new Set([]);
+    for (const option of program.options) {
+      const hasValue = option.flags.includes("[");
+      if (option.short) {
+        babelOptions.add(option.short);
+        if (hasValue) babelOptionsWithValue.add(option.short);
+      }
+      if (option.long) {
+        babelOptions.add(option.long);
+        if (hasValue) babelOptionsWithValue.add(option.long);
+      }
+    }
+
+    const { programArgs, fileName, userArgs, explicitSeparator } = splitArgs(
+      process.argv.slice(2),
+      babelOptionsWithValue,
     );
 
-    const v8FlagsSet = new Set(v8Flags);
+    const babelArgs: string[] = [];
+    const nodeArgs: string[] = [];
+
+    for (let i = 0; i < programArgs.length; i++) {
+      const arg = programArgs[i];
+      const list = babelOptions.has(arg.split("=")[0]) ? babelArgs : nodeArgs;
+      list.push(arg);
+      if (i + 1 < programArgs.length && programArgs[i + 1][0] !== "-") {
+        list.push(programArgs[++i]);
+      }
+    }
+
+    if (!explicitSeparator) {
+      const isV8flag = await getV8Flags();
+
+      const ambiguousArgsNames: string[] = [];
+      const ambiguousArgs: string[] = [];
+      let unambiguousArgs: string[] | null = null;
+      for (let i = 0; i < userArgs.length; i++) {
+        const [arg, value] = userArgs[i].split("=");
+        if (babelOptions.has(arg) || isV8flag(arg)) {
+          unambiguousArgs ??= userArgs.slice(0, i);
+          ambiguousArgsNames.push(arg);
+          ambiguousArgs.push(userArgs[i]);
+          if (
+            value === undefined &&
+            (babelOptionsWithValue.has(arg) || nodeFlagsWithValue.has(arg)) &&
+            i + 1 < userArgs.length
+          ) {
+            ambiguousArgs.push(userArgs[++i]);
+          }
+        } else {
+          unambiguousArgs?.push(userArgs[i]);
+        }
+      }
+      if (ambiguousArgsNames.length > 0) {
+        const them = ambiguousArgsNames.length === 1 ? "it" : "them";
+        const they = ambiguousArgsNames.length === 1 ? "it" : "they";
+        const are = ambiguousArgsNames.length === 1 ? "is" : "are";
+        console.warn(
+          `Warning: ${ambiguousArgsNames.join(", ")} ${are} a valid option for Babel or Node.js, ` +
+            `but ${they} ${are} defined after the script name. Up to Babel 7 ${they} would have ` +
+            `been passed to Babel, while now ${they} ${are} passed to the script itself.\n` +
+            `  If this is intended, you can silence this warning by explicitly ` +
+            `passing the -- separator before the script name:\n` +
+            `    babel-node ${programArgs.join(" ")} -- ${fileName} ${userArgs.join(" ")}\n` +
+            `  If the intention is to pass ${them} to Babel, move ${them} before the filename:\n` +
+            `    babel-node ${programArgs.join(" ")} ${ambiguousArgs.join(" ")} ${fileName} ${unambiguousArgs.join(" ")}\n`,
```

Comment:
Nit: Can we swap these two? I assume most people are intended to pass options to Babel / Node so let's offer the majority choice first.